File: app/desktop/src/features/cloud/providerLogin.ts
import { cloudApiBaseUrl } from './cloudApiEnvironment';
import { cloudFetchImpl } from './cloudTransport';
import { isOmpUnavailableResponse, OMP_UNAVAILABLE_MESSAGE } from './ompAvailability';
import { publishableAccountLabel } from './routeAccountChoice';
import { loadSession } from './session';

// Hosted OMP sign-in sessions. The server runs OMP's own login flow for a
// provider and exposes each step; the desktop renders the step and relays
// user input. Credentials never pass through this client except the API key
// or pasted code the user types, which goes straight to the session input.

export type OmpLoginKind = 'api-key' | 'oauth-code' | 'device-code' | 'custom' | 'env-only';

/** Per-provider login description from the pinned OMP catalog. */
export type OmpLoginSpec = {
  kind: OmpLoginKind;
  name: string;
  instructions: string | null;
  prompt: string | null;
  placeholder: string | null;
  authUrl: string | null;
  validates: boolean;
  pasteKey: boolean;
  manualOnly: boolean;
  callbackPort: number | null;
  /** Path of OMP's loopback redirect, for example "/auth/callback"; older catalogs omit it. */
  callbackPath?: string | null;
  hook: string | null;
  apiKeyFormat: string | null;
  envVars: string[];
  storeCredentialsAs: string | null;
  /** True when the provider also takes an API key (paste key or a *_API_KEY variable). */
  acceptsApiKeyMethod?: boolean;
};

export type ProviderLoginAuthLink = { url: string; launchUrl?: string | null; instructions?: string | null };

export type ProviderLoginStep =
  | { type: 'open-url'; url: string; launchUrl?: string | null; instructions?: string | null }
  | { type: 'prompt'; message: string; placeholder?: string | null; secret?: boolean; allowEmpty?: boolean }
  | { type: 'paste-code'; instructions?: string | null }
  | { type: 'progress'; message: string }
  | { type: 'api-key'; instructions?: string | null; placeholder?: string | null; authUrl?: string | null };

export type ProviderLoginStatus = 'running' | 'awaiting-input' | 'claiming' | 'completed' | 'failed' | 'cancelled';

const terminalStatuses: ReadonlySet<ProviderLoginStatus> = new Set<ProviderLoginStatus>(['completed', 'failed', 'cancelled']);

/** Completed, failed and cancelled end a sign-in; the server's answer for them is final. */
export function isTerminalLoginStatus(status: ProviderLoginStatus) {
  return terminalStatuses.has(status);
}

export type ProviderLoginSnapshot = { snapshotId: string; provider: string; authChoice: string; label: string };

export type ProviderLoginSession = {
  sessionId: string;
  status: ProviderLoginStatus;
  step: ProviderLoginStep | null;
  auth: ProviderLoginAuthLink | null;
  /** Pass back as `after` when long-polling; null for states the server records itself (claiming, completed, cancelled). */
  version: number | null;
  error?: string | null;
  reason?: string | null;
  snapshot?: ProviderLoginSnapshot | null;
};

export type ProviderLoginStartInput = {
  provider: string;
  label: string;
  /** `device` selects a device-code hook such as openai-codex-device. */
  mode?: 'device';
  /** `api-key` asks OMP for key entry instead of the provider's default sign-in. */
  method?: 'default' | 'api-key';
};

export type ProviderLoginErrorCode =
  | 'login_not_found'
  | 'login_expired'
  | 'login_failed'
  | 'login_unsupported'
  | 'invalid_login_input'
  | 'rate_limited'
  | 'omp_unavailable'
  | 'provider_auth_not_configured'
  | 'omp_busy'
  | 'not_signed_in'
  | 'network_error'
  | 'unknown';

const knownErrorCodes = new Set<ProviderLoginErrorCode>([
  'login_not_found', 'login_expired', 'login_failed', 'login_unsupported', 'invalid_login_input',
  'rate_limited', 'omp_unavailable', 'provider_auth_not_configured', 'omp_busy', 'not_signed_in', 'network_error',
]);

/** Failures worth retrying in place: OMP restarting, or a dropped connection. */
export function isTransientLoginError(error: ProviderLoginError) {
  return error.code === 'omp_unavailable' || error.code === 'network_error';
}

function failureReason(reason?: string | null) {
  const text = reason?.trim().replace(/_/g, ' ');
  return text ? ` (${text.length > 160 ? `${text.slice(0, 157)}…` : text})` : '';
}

/** Recovery text for each server error code; distinct so the user knows what to do next. */
export function providerLoginErrorMessage(code: ProviderLoginErrorCode, reason?: string | null): string {
  switch (code) {
    case 'login_not_found': return 'This sign-in is no longer available. Start again.';
    case 'login_expired': return 'This sign-in expired before it finished. Start again.';
    case 'login_failed': return `The provider did not complete sign-in${failureReason(reason)}. Try again.`;
    case 'login_unsupported': return 'OMP cannot add this account from Kordi yet.';
    case 'invalid_login_input': return 'OMP did not accept that value. Check it and try again.';
    case 'rate_limited': return 'Too many sign-in attempts. Wait a minute and try again.';
    case 'omp_unavailable': return 'OMP is not reachable right now. Try again in a moment.';
    case 'provider_auth_not_configured': return OMP_UNAVAILABLE_MESSAGE;
    case 'omp_busy': return 'OMP is busy with other sign-ins. Try again in a moment.';
    case 'not_signed_in': return 'Sign in to Kordi before adding an account.';
    case 'network_error': return 'Kordi could not reach the server. Check your connection and try again.';
    default: return 'Sign-in could not continue. Try again.';
  }
}

export class ProviderLoginError extends Error {
  readonly code: ProviderLoginErrorCode;
  readonly status: number;
  readonly reason: string | null;
  /** The backend has no OMP sign-in (an older server, or one not configured for it); see ompAvailability.ts. */
  readonly ompUnavailable: boolean;

  constructor(code: ProviderLoginErrorCode, status = 0, reason: string | null = null) {
    super(providerLoginErrorMessage(code, reason));
    this.name = 'ProviderLoginError';
    this.code = code;
    this.status = status;
    this.reason = reason;
    this.ompUnavailable = code === 'provider_auth_not_configured';
  }
}

export function toProviderLoginError(caught: unknown): ProviderLoginError {
  if (caught instanceof ProviderLoginError) return caught;
  return new ProviderLoginError('unknown');
}

/** Maps an HTTP error body ({ errorCode | error.code, reason }) to a typed error. */
export function providerLoginErrorFromResponse(status: number, body: unknown): ProviderLoginError {
  const record = (body && typeof body === 'object' ? body : {}) as Record<string, unknown>;
  const nested = (record.error && typeof record.error === 'object' ? record.error : {}) as Record<string, unknown>;
  const rawCode = [record.errorCode, record.code, nested.code].find((value): value is string => typeof value === 'string');
  const reason = [record.reason, nested.reason].find((value): value is string => typeof value === 'string');
  if (isOmpUnavailableResponse(status, rawCode)) return new ProviderLoginError('provider_auth_not_configured', status, reason ?? null);
  // A 503 names its cause; an unknown one is not taken to mean OMP is missing.
  const code = rawCode && knownErrorCodes.has(rawCode as ProviderLoginErrorCode)
    ? rawCode as ProviderLoginErrorCode
    : status === 401 ? 'not_signed_in'
      : status === 410 ? 'login_expired'
        : status === 429 ? 'rate_limited'
          : 'unknown';
  return new ProviderLoginError(code, status, reason ?? null);
}

export interface ProviderLoginClient {
  start(input: ProviderLoginStartInput, signal?: AbortSignal): Promise<ProviderLoginSession>;
  /** Long-polls until the session version passes `after` or the server wait ends. */
  poll(sessionId: string, after: number, signal?: AbortSignal): Promise<ProviderLoginSession>;
  submit(sessionId: string, value: string, signal?: AbortSignal): Promise<ProviderLoginSession | null>;
  /** Resolves with the session's final state when the server sends one, for example a sign-in that completed first. */
  cancel(sessionId: string): Promise<ProviderLoginSession | null>;
}

const loginPath = '/v1/cloud/agent-provider-auth/login';

function isSession(value: unknown): value is ProviderLoginSession {
  return Boolean(value && typeof value === 'object' && typeof (value as ProviderLoginSession).sessionId === 'string');
}

export function createCloudProviderLoginClient(options: {
  baseUrl?: () => string;
  fetchImpl?: typeof fetch;
  token?: () => Promise<string | null>;
  pollWaitSeconds?: number;
} = {}): ProviderLoginClient {
  const baseUrl = options.baseUrl ?? (() => cloudApiBaseUrl());
  const token = options.token ?? (async () => (await loadSession())?.token ?? null);
  const waitSeconds = options.pollWaitSeconds ?? 25;

  const request = async (path: string, init: RequestInit, timeoutMs: number): Promise<unknown> => {
    const bearer = await token();
    if (!bearer) throw new ProviderLoginError('not_signed_in', 401);
    const fetchImpl = options.fetchImpl ?? cloudFetchImpl();
    const timeout = new AbortController();
    const timer = setTimeout(() => timeout.abort(), timeoutMs);
    const abortFromCaller = () => timeout.abort();
    init.signal?.addEventListener('abort', abortFromCaller, { once: true });
    let response: Response;
    try {
      response = await fetchImpl(`${baseUrl()}${loginPath}${path}`, {
        ...init,
        signal: timeout.signal,
        headers: { ...(init.body ? { 'content-type': 'application/json' } : {}), authorization: `Bearer ${bearer}` },
      });
    } catch (caught) {
      if (init.signal?.aborted) throw caught;
      throw new ProviderLoginError('network_error');
    } finally {
      clearTimeout(timer);
      init.signal?.removeEventListener('abort', abortFromCaller);
    }
    const text = response.status === 204 ? '' : await response.text();
    let body: unknown = null;
    try { body = text ? JSON.parse(text) as unknown : null; } catch { body = null; }
    if (!response.ok) throw providerLoginErrorFromResponse(response.status, body);
    return body;
  };

  return {
    async start(input, signal) {
      const payload = { ...input, label: publishableAccountLabel(input.label) };
      const body = await request('/start', { method: 'POST', body: JSON.stringify(payload), signal }, 30_000);
      if (!isSession(body)) throw new ProviderLoginError('unknown');
      return body;
    },
    async poll(sessionId, after, signal) {
      const query = `?wait=${waitSeconds}&after=${after}`;
      const body = await request(`/${encodeURIComponent(sessionId)}${query}`, { method: 'GET', signal }, (waitSeconds + 15) * 1000);
      if (!isSession(body)) throw new ProviderLoginError('unknown');
      return body;
    },
    async submit(sessionId, value, signal) {
      const body = await request(`/${encodeURIComponent(sessionId)}/input`, { method: 'POST', body: JSON.stringify({ value }), signal }, 30_000);
      return isSession(body) ? body : null;
    },
    async cancel(sessionId) {
      const body = await request(`/${encodeURIComponent(sessionId)}/cancel`, { method: 'POST', body: '{}' }, 15_000);
      return isSession(body) ? body : null;
    },
  };
}

/** Finds the device code in OMP instructions such as "Enter code: ABCD-1234". */
export function parseLoginUserCode(text?: string | null): string | null {
  if (!text) return null;
  const labelled = text.match(/\bcode\b[^A-Za-z0-9]{0,4}([A-Z0-9]{3,}(?:-[A-Z0-9]{3,})*)/i);
  if (labelled && /\d|-/.test(labelled[1])) return labelled[1];
  return text.match(/\b([A-Z0-9]{4,}-[A-Z0-9]{4,})\b/)?.[1] ?? null;
}

/** Paste-code guidance when OMP's step carries none. */
export const pasteCodeFallback = 'Paste the code or the full redirect URL from the browser.';
/** Transcript line once the desktop receives the provider's localhost redirect. */
export const loginCallbackReceivedText = 'Sign-in received from the browser';
/** Shown under the paste field when the desktop cannot listen for the redirect. */
export const loginCallbackFallbackHint = "After you approve in the browser it will land on a localhost page that cannot load. Copy that page's full address and paste it here.";

/** Shown when another program on this Mac holds the provider's loopback port. */
export function loginCallbackPortBusyHint(port: number) {
  return `Port ${port} is in use on this Mac. Close the program using it, or paste the callback link below.`;
}

export type ProviderLoginInputView = {
  kind: 'api-key' | 'prompt' | 'paste-code';
  message: string | null;
  placeholder: string | null;
  secret: boolean;
  allowEmpty: boolean;
  authUrl: string | null;
};

/** Lines kept on the login page like OMP's dialog: answers (secrets hidden) and progress. */
export type ProviderLoginTranscriptEntry =
  | { type: 'answer'; label: string; value: string | null }
  | { type: 'progress'; text: string };

export type ProviderLoginView = {
  phase: 'idle' | 'starting' | 'waiting' | 'input' | 'submitting' | 'completed' | 'failed' | 'cancelled';
  transcript: ProviderLoginTranscriptEntry[];
  sessionId: string | null;
  version: number;
  auth: ProviderLoginAuthLink | null;
  userCode: string | null;
  statusLine: string | null;
  input: ProviderLoginInputView | null;
  error: { code: ProviderLoginErrorCode; message: string } | null;
  snapshot: ProviderLoginSnapshot | null;
  /** Desktop capture of the provider's localhost redirect for a browser sign-in. */
  callback: 'off' | 'listening' | 'unavailable' | 'port-busy' | 'received';
  /** The loopback port another program holds, while `callback` is `port-busy`. */
  callbackPort: number | null;
};

export type ProviderLoginAction =
  | { type: 'start' }
  | { type: 'session'; session: ProviderLoginSession }
  | { type: 'submit'; answer?: { label: string; value: string; secret: boolean } }
  | { type: 'error'; error: ProviderLoginError }
  | { type: 'cancelled' }
  | { type: 'callback'; state: 'listening' | 'unavailable' | 'received' }
  | { type: 'callback'; state: 'port-busy'; port: number }
  | { type: 'reset' };

export const initialProviderLoginView: ProviderLoginView = {
  phase: 'idle', transcript: [], sessionId: null, version: -1, auth: null, userCode: null,
  statusLine: null, input: null, error: null, snapshot: null, callback: 'off', callbackPort: null,
};

function withProgress(transcript: ProviderLoginTranscriptEntry[], text: string | null | undefined) {
  if (!text) return transcript;
  const last = [...transcript].reverse().find((entry) => entry.type === 'progress');
  return last?.type === 'progress' && last.text === text ? transcript : [...transcript, { type: 'progress' as const, text }];
}

function inputForStep(step: ProviderLoginStep): ProviderLoginInputView | null {
  switch (step.type) {
    case 'api-key':
      return { kind: 'api-key', message: step.instructions ?? null, placeholder: step.placeholder ?? null, secret: true, allowEmpty: false, authUrl: step.authUrl ?? null };
    case 'prompt':
      return { kind: 'prompt', message: step.message, placeholder: step.placeholder ?? null, secret: Boolean(step.secret), allowEmpty: Boolean(step.allowEmpty), authUrl: null };
    case 'paste-code':
      return { kind: 'paste-code', message: step.instructions ?? pasteCodeFallback, placeholder: null, secret: false, allowEmpty: false, authUrl: null };
    default:
      return null;
  }
}

function viewFromSession(state: ProviderLoginView, session: ProviderLoginSession): ProviderLoginView {
  const sameSession = state.sessionId === session.sessionId;
  // A saved account stays saved: nothing later for this sign-in replaces the result.
  if (sameSession && state.phase === 'completed') return state;
  // The server sends a null version for states it records itself, so versions compare only as numbers.
  const version = typeof session.version === 'number' ? session.version : null;
  if (sameSession && !isTerminalLoginStatus(session.status)) {
    if (state.phase === 'cancelled') return state;
    // An older state, or the unchanged one a long poll returns when its wait ends: nothing new,
    // so an error or a submit in progress stays on screen.
    if (version !== null && version <= state.version) return state;
  }
  const auth = session.auth ?? (session.step?.type === 'open-url'
    ? { url: session.step.url, launchUrl: session.step.launchUrl, instructions: session.step.instructions }
    : sameSession ? state.auth : null);
  const base = {
    ...state,
    sessionId: session.sessionId,
    version: version ?? (sameSession ? state.version : -1),
    auth,
    userCode: parseLoginUserCode(auth?.instructions) ?? (sameSession ? state.userCode : null),
    error: null,
  };
  if (session.status === 'completed') {
    return { ...base, phase: 'completed', input: null, statusLine: null, snapshot: session.snapshot ?? null };
  }
  if (session.status === 'failed') {
    const code = (session.error && knownErrorCodes.has(session.error as ProviderLoginErrorCode)
      ? session.error : 'login_failed') as ProviderLoginErrorCode;
    return { ...base, phase: 'failed', input: null, statusLine: null, error: { code, message: providerLoginErrorMessage(code, session.reason) } };
  }
  if (session.status === 'cancelled') return { ...base, phase: 'cancelled', input: null, statusLine: null };

  const step = session.step;
  const input = step ? inputForStep(step) : null;
  if (step?.type === 'progress') base.transcript = withProgress(base.transcript, step.message);
  if (input && session.status === 'awaiting-input') {
    return { ...base, phase: 'input', input, statusLine: null };
  }
  return {
    ...base,
    phase: 'waiting',
    input: null,
    statusLine: step?.type === 'progress' ? step.message
      : step?.type === 'open-url' ? step.instructions ?? null
        : input ? null : base.statusLine,
  };
}

// A prompt that asks for the authorization code or redirect URL rather than another detail.
const codePromptPattern = /\b(code|url|redirect)\b/i;

/** A step that takes the captured redirect: OMP's paste-code step, or a prompt for the code or URL. */
export function acceptsLoginCallback(session: Pick<ProviderLoginSession, 'status' | 'step'>) {
  const step = session.step;
  if (session.status !== 'awaiting-input' || !step) return false;
  return step.type === 'paste-code' || (step.type === 'prompt' && codePromptPattern.test(step.message));
}

/**
 * The hint when the redirect cannot be captured. A busy port is explained as
 * soon as it is known and beside a paste step; otherwise the hint goes under
 * the paste field only.
 */
export function loginCallbackHint(view: ProviderLoginView): string | null {
  const pasteStep = view.input
    ? view.input.kind === 'paste-code' || (view.input.kind === 'prompt' && codePromptPattern.test(view.input.message ?? ''))
    : null;
  if (view.callback === 'port-busy' && view.callbackPort !== null && pasteStep !== false) {
    return loginCallbackPortBusyHint(view.callbackPort);
  }
  return view.callback === 'unavailable' && pasteStep ? loginCallbackFallbackHint : null;
}

/** Maps hosted sign-in session updates to what the add-account row shows. */
export function providerLoginReducer(state: ProviderLoginView, action: ProviderLoginAction): ProviderLoginView {
  switch (action.type) {
    case 'start':
      return { ...initialProviderLoginView, phase: 'starting', statusLine: 'Starting sign-in with OMP…' };
    case 'session':
      return viewFromSession(state, action.session);
    case 'submit': {
      const statusLine = state.input?.kind === 'api-key' ? 'Verifying with OMP…' : 'Sending to OMP…';
      const answered = action.answer
        ? [...state.transcript, { type: 'answer' as const, label: action.answer.label, value: action.answer.secret ? null : action.answer.value }]
        : state.transcript;
      return {
        ...state,
        phase: 'submitting',
        error: null,
        statusLine,
        transcript: state.input?.kind === 'api-key' ? withProgress(answered, statusLine) : answered,
      };
    }
    case 'error':
      // A late failure, for example of a submit that raced the result, never hides a saved account.
      if (state.phase === 'completed') return state;
      if (action.error.code === 'invalid_login_input' && state.input) {
        return { ...state, phase: 'input', statusLine: null, error: { code: action.error.code, message: action.error.message } };
      }
      return { ...state, phase: 'failed', input: null, statusLine: null, error: { code: action.error.code, message: action.error.message } };
    case 'cancelled':
      if (state.phase === 'completed') return state;
      return { ...state, phase: 'cancelled', input: null, statusLine: null };
    case 'callback':
      // The redirect itself carries the code and never enters the view.
      if (action.state === 'received') {
        return { ...state, callback: 'received', callbackPort: null, transcript: withProgress(state.transcript, loginCallbackReceivedText) };
      }
      return { ...state, callback: action.state, callbackPort: action.state === 'port-busy' ? action.port : null };
    case 'reset':
      return initialProviderLoginView;
  }
}

File: app/ios/Kordi/Core/Models/ProviderLogin.swift
import Foundation

// MARK: - Login session wire format

enum ProviderLoginStatus: String, Decodable, Equatable {
    case running
    case awaitingInput = "awaiting-input"
    /// The server is saving the finished login; OMP needs nothing more.
    case claiming
    case completed
    case failed
    case cancelled

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        // An unknown status must stop polling rather than loop forever.
        self = ProviderLoginStatus(rawValue: raw) ?? .failed
    }

    var isTerminal: Bool { self == .completed || self == .failed || self == .cancelled }
}

enum ProviderLoginStep: Decodable, Equatable {
    case openURL(url: String, launchURL: String?, instructions: String?)
    case prompt(message: String, placeholder: String?, secret: Bool, allowEmpty: Bool)
    case pasteCode(instructions: String?)
    case progress(message: String)
    case apiKey(instructions: String?, placeholder: String?, authURL: String?)
    case unsupported(type: String)

    /// Steps that ask the user for a value.
    var isInput: Bool {
        switch self {
        case .prompt, .pasteCode, .apiKey: true
        case .openURL, .progress, .unsupported: false
        }
    }

    private enum CodingKeys: String, CodingKey {
        case type, url, launchUrl, instructions, message, placeholder, secret, allowEmpty, authUrl
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let type = try values.decode(String.self, forKey: .type)
        switch type {
        case "open-url":
            self = .openURL(
                url: try values.decode(String.self, forKey: .url),
                launchURL: try values.decodeIfPresent(String.self, forKey: .launchUrl),
                instructions: try values.decodeIfPresent(String.self, forKey: .instructions)
            )
        case "prompt":
            self = .prompt(
                message: try values.decodeIfPresent(String.self, forKey: .message) ?? "",
                placeholder: try values.decodeIfPresent(String.self, forKey: .placeholder),
                secret: try values.decodeIfPresent(Bool.self, forKey: .secret) ?? false,
                allowEmpty: try values.decodeIfPresent(Bool.self, forKey: .allowEmpty) ?? false
            )
        case "paste-code":
            self = .pasteCode(instructions: try values.decodeIfPresent(String.self, forKey: .instructions))
        case "progress":
            self = .progress(message: try values.decodeIfPresent(String.self, forKey: .message) ?? "")
        case "api-key":
            self = .apiKey(
                instructions: try values.decodeIfPresent(String.self, forKey: .instructions),
                placeholder: try values.decodeIfPresent(String.self, forKey: .placeholder),
                authURL: try values.decodeIfPresent(String.self, forKey: .authUrl)
            )
        default:
            self = .unsupported(type: type)
        }
    }
}

/// The latest sign-in link. The server keeps it after later steps so the page
/// and any device code stay available.
struct ProviderLoginAuth: Decodable, Equatable {
    let url: String
    let launchUrl: String?
    let instructions: String?

    init(url: String, launchUrl: String? = nil, instructions: String? = nil) {
        self.url = url
        self.launchUrl = launchUrl
        self.instructions = instructions
    }
}

struct ProviderLoginSnapshot: Decodable, Equatable {
    let snapshotId: String
    let provider: String
    let authChoice: String
    let label: String?
}

struct ProviderLoginSession: Decodable, Equatable {
    let sessionId: String?
    let status: ProviderLoginStatus
    let step: ProviderLoginStep?
    let auth: ProviderLoginAuth?
    let error: String?
    let reason: String?
    /// Changes on every update; sent back as `after` so a poll returns at once.
    let version: Int?
    let snapshot: ProviderLoginSnapshot?

    init(
        sessionId: String? = nil,
        status: ProviderLoginStatus,
        step: ProviderLoginStep? = nil,
        auth: ProviderLoginAuth? = nil,
        error: String? = nil,
        reason: String? = nil,
        version: Int? = nil,
        snapshot: ProviderLoginSnapshot? = nil
    ) {
        self.sessionId = sessionId
        self.status = status
        self.step = step
        self.auth = auth
        self.error = error
        self.reason = reason
        self.version = version
        self.snapshot = snapshot
    }
}

// MARK: - View state

enum ProviderLoginEvent: Equatable {
    case start
    case session(ProviderLoginSession)
    case submit
    case failure(code: String, reason: String?)
    case cancel
    case reset
}

/// What the Add-account section shows. `ProviderLoginViewState.reduce` is the
/// only place that maps login-session steps and errors to this state.
struct ProviderLoginViewState: Equatable {
    enum Phase: Equatable {
        case idle, starting, signIn, input, working, completed, failed, cancelled
    }

    enum InputKind: Equatable {
        case apiKey, pasteCode, prompt
    }

    var phase: Phase = .idle
    var inputKind: InputKind?
    var instructions: String?
    var message: String?
    var placeholder: String?
    var isSecret = false
    var allowsEmpty = false
    /// The last sign-in page OMP asked the user to open; kept across later steps.
    var signInURL: URL?
    var keyURL: URL?
    /// A device code parsed from OMP's instructions or progress text.
    var userCode: String?
    /// Instructions that came with the sign-in link.
    var authInstructions: String?
    var status: String?
    var inputError: String?
    var failure: String?
    var snapshot: ProviderLoginSnapshot?

    var isActive: Bool {
        switch phase {
        case .starting, .signIn, .input, .working: true
        default: false
        }
    }

    /// Stable step name for accessibility and tests.
    var stepName: String {
        switch phase {
        case .idle: "idle"
        case .starting: "starting"
        case .signIn: userCode == nil ? "open-url" : "device-code"
        case .input:
            switch inputKind {
            case .apiKey: "api-key"
            case .pasteCode: "paste-code"
            case .prompt, nil: "prompt"
            }
        case .working: "progress"
        case .completed: "completed"
        case .failed: "failed"
        case .cancelled: "cancelled"
        }
    }

    static let verifyingMessage = "Verifying with OMP…"
    static let savingMessage = "Saving the account…"

    static func reduce(_ state: ProviderLoginViewState, _ event: ProviderLoginEvent) -> ProviderLoginViewState {
        switch event {
        case .start:
            return ProviderLoginViewState(phase: .starting, status: "Connecting to OMP…")
        case .reset:
            return ProviderLoginViewState()
        case .cancel:
            return ProviderLoginViewState(phase: .cancelled)
        case .submit:
            var next = state
            next.phase = .working
            next.status = verifyingMessage
            next.inputError = nil
            return next
        case .failure(let code, let reason):
            if code == "invalid_login_input", state.inputKind != nil,
               state.phase == .working || state.phase == .input {
                var next = state
                next.phase = .input
                next.status = nil
                next.inputError = failureMessage(code: code, reason: reason)
                return next
            }
            return failed(state, failureMessage(code: code, reason: reason))
        case .session(let session):
            return apply(session, to: state)
        }
    }

    private static func apply(_ session: ProviderLoginSession, to state: ProviderLoginViewState) -> ProviderLoginViewState {
        switch session.status {
        case .completed:
            var next = ProviderLoginViewState(phase: .completed)
            next.snapshot = session.snapshot
            return next
        case .cancelled:
            return ProviderLoginViewState(phase: .cancelled)
        case .failed:
            // OMP reports its failure classification in `error`.
            return failed(state, failureMessage(code: "login_failed", reason: session.reason ?? session.error))
        case .running, .awaitingInput, .claiming:
            break
        }
        var next = state
        next.inputError = nil
        next.failure = nil
        // `auth` outlives the step that created it; keep its page and code visible.
        if let auth = session.auth {
            guard let target = httpsURL(auth.launchUrl) ?? httpsURL(auth.url) else {
                return failed(state, "OMP returned a sign-in address Kordi cannot open.")
            }
            next.signInURL = target
            next.userCode = userCode(in: auth.instructions) ?? next.userCode
            next.authInstructions = auth.instructions?.nonEmpty ?? next.authInstructions
        }
        guard let step = session.step else {
            if next.phase != .signIn { next.phase = .working }
            next.status = session.status == .claiming ? savingMessage : next.status ?? "Waiting for OMP…"
            return next
        }
        if step.isInput, session.status != .awaitingInput {
            // OMP reads input only while awaiting it; an input step on a running
            // session is the value it is still checking.
            next.phase = .working
            next.status = state.phase == .working ? state.status ?? verifyingMessage : "Waiting for OMP…"
            return next
        }
        switch step {
        case .openURL(let url, let launchURL, let instructions):
            guard let target = httpsURL(launchURL) ?? httpsURL(url) else {
                return failed(state, "OMP returned a sign-in address Kordi cannot open.")
            }
            next.phase = .signIn
            next.inputKind = nil
            next.signInURL = target
            next.instructions = instructions?.nonEmpty
            next.authInstructions = instructions?.nonEmpty ?? next.authInstructions
            next.userCode = userCode(in: instructions) ?? next.userCode
            next.status = nil
        case .prompt(let message, let placeholder, let secret, let allowEmpty):
            next.phase = .input
            next.inputKind = .prompt
            next.message = message.nonEmpty
            next.placeholder = placeholder?.nonEmpty
            next.isSecret = secret
            next.allowsEmpty = allowEmpty
            next.status = nil
        case .pasteCode(let instructions):
            next.phase = .input
            next.inputKind = .pasteCode
            next.instructions = instructions?.nonEmpty ?? state.instructions
            next.message = nil
            next.placeholder = nil
            next.isSecret = false
            next.allowsEmpty = false
            next.status = nil
        case .progress(let message):
            if next.phase != .signIn { next.phase = .working }
            next.status = message.nonEmpty ?? "Waiting for OMP…"
            next.userCode = next.userCode ?? userCode(in: message)
        case .apiKey(let instructions, let placeholder, let authURL):
            next.phase = .input
            next.inputKind = .apiKey
            next.instructions = instructions?.nonEmpty
            next.message = nil
            next.placeholder = placeholder?.nonEmpty
            next.keyURL = httpsURL(authURL)
            next.isSecret = true
            next.allowsEmpty = false
            next.status = nil
        case .unsupported:
            return failed(state, "This sign-in step needs Kordi on your Mac. Open Settings → Authentication there.")
        }
        return next
    }

    private static func failed(_ state: ProviderLoginViewState, _ message: String) -> ProviderLoginViewState {
        var next = ProviderLoginViewState(phase: .failed)
        next.failure = message
        return next
    }

    static func failureMessage(code: String, reason: String?) -> String {
        switch code {
        case "login_not_found":
            return "This sign-in is no longer available. Start again."
        case "login_expired":
            return "This sign-in expired. Start again."
        case "login_failed":
            return reasonMessage(reason) ?? "OMP could not finish signing in. Try again."
        case "invalid_login_input":
            return "OMP did not accept that value. Check it and try again."
        case "login_unsupported":
            return unsupportedMessage
        case "omp_busy":
            return "OMP is handling other sign-ins. Wait a moment, then try again."
        case "rate_limited":
            return "Too many sign-in attempts. Wait a minute, then try again."
        case OMPBackendSupport.notConfiguredCode:
            return OMPBackendSupport.unavailableMessage
        case OMPBackendSupport.transientCode:
            return OMPBackendSupport.transientMessage
        case "network_error", "proxy_unreachable":
            return "Could not reach Kordi Cloud. Check your connection and try again."
        case "invalid_session", "account_missing":
            return "Your Kordi session expired. Sign in again, then add the account."
        case "login_cancelled":
            return "Sign-in was canceled."
        case "provider_auth_error":
            return "Kordi Cloud could not save this account. Try again."
        default:
            return reasonMessage(reason) ?? "Sign-in failed. Try again."
        }
    }

    private static let unsupportedMessage = "This sign-in method is not available from Kordi Cloud yet. Add the account in Kordi on your Mac."

    /// The failure classifications OMP and Kordi Cloud report. Unknown values
    /// are never shown: they are identifiers, not sentences.
    static func reasonMessage(_ reason: String?) -> String? {
        switch reason?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "invalid_input":
            return "OMP did not accept that value. Check it and start again."
        case "provider_rejected":
            return "The provider did not accept this sign-in. Check the account and try again."
        case "timeout":
            return "The sign-in timed out. Start again."
        case "unsupported_flow", "unknown_provider":
            return unsupportedMessage
        case "session_lost", "start_failed":
            return "OMP stopped this sign-in. Start again."
        case "claim_failed", "save_failed":
            return "Sign-in finished, but Kordi Cloud could not save the account. Start again."
        default:
            return nil
        }
    }

    static func httpsURL(_ value: String?) -> URL? {
        guard let value = value?.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty,
              let url = URL(string: value), url.scheme?.lowercased() == "https", url.host != nil else {
            return nil
        }
        return url
    }

    /// Finds a device code such as `ABCD-1234` in OMP text like `Enter code: ABCD-1234`.
    static func userCode(in text: String?) -> String? {
        guard let text else { return nil }
        let patterns = [
            #"(?i)code:\s*([A-Z0-9]{3,}(?:-[A-Z0-9]{3,})*)"#,
            #"\b([A-Z0-9]{3,}(?:-[A-Z0-9]{3,})+)\b"#,
        ]
        for pattern in patterns {
            guard let regex = try? NSRegularExpression(pattern: pattern),
                  let match = regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
                  let range = Range(match.range(at: 1), in: text) else { continue }
            return String(text[range])
        }
        return nil
    }
}
